cmd/batch-expire.go

```diff
@@ -534,18 +535,34 @@ func (r *BatchJobExpire) Start(ctx context.Context, api ObjectLayer, job BatchJo
 		return err
 	}
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	ctx, cancelCancel := context.WithCancelCause(ctx)
+	defer func() {
+		if context.Cause(ctx) == nil {
+			cancelCancel(nil)
+		}
+	}()
 
 	results := make(chan itemOrErr[ObjectInfo], workerSize)
-	if err := api.Walk(ctx, r.Bucket, r.Prefix, results, WalkOptions{
-		Marker:       lastObject,
-		LatestOnly:   false, // we need to visit all versions of the object to implement purge: retainVersions
-		VersionsSort: WalkVersionsSortDesc,
-	}); err != nil {
-		// Do not need to retry if we can't list objects on source.
-		return err
-	}
+	go func() {
+		prefixes := strings.Split(r.Prefix, ",")
+		for _, prefix := range prefixes {
+			prefixResultCh := make(chan itemOrErr[ObjectInfo], workerSize)
+			err := api.Walk(ctx, r.Bucket, strings.TrimSpace(prefix), prefixResultCh, WalkOptions{
+				Marker:       lastObject,
+				LatestOnly:   false, // we need to visit all versions of the object to implement purge: retainVersions
+				VersionsSort: WalkVersionsSortDesc,
+			})
+			if err != nil {
+				xioutil.SafeClose(results)
+				cancelCancel(err)
+				return
+			}
+			for result := range prefixResultCh {
+				results <- result
+			}
+		}
+		xioutil.SafeClose(results)
+	}()
```

Comment:
Below you need to look at

```
			case <-ctx.Done():
				// persist in-memory state immediately before exiting due to context cancellation.
				batchLogIf(ctx, ri.updateAfter(ctx, api, 0, job))
```
A canceled config is being sent to the saveConfig. Pretty sure that does nothing.